### PR TITLE
Add documentation of demos

### PR DESCRIPTION
**Pull request by OscarShiang**
Add a document into docs/

Although I am still working on the README files of the demos, I think I will just put the links here.
I hope that we can save some time so that we don't have to revise this file whenever we add new README files :)

### CODE REVIEW

**Comment by OscarShiang** [comment]
By the way, should we put some brief description after each demo link?

File: docs/demos.md [hunk 1]
```diff
@@ -0,0 +1,29 @@
+# AMT Demos 
+
+These documentations give an overview of how the demos work and explain some mechanism in these demos. (including some functions we use in the demo)
```
Comment (DavidSpickett) [inline docs/demos.md]: "The documents listed below give"
"explain some mechanisms in these"

You can remove "(including some functions we use in the demo)" I think it's implied already.

File: docs/demos.md [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+For a better understanding, we suggest that you can try to revise some code to change and verify the behaviour of the demos or dig into them by using commands like `make debug` and trace them one by one.
```
Comment (DavidSpickett) [inline docs/demos.md]: I like this but the format is a bit awkward. I'd do something like this to be clearer:

For the best understanding we suggest that you:
* read the document and demo source side by side
* make some changes to the demo to verify its behaviour
* follow it in a debugger with `make debug_<demo>`

Btw I hadn't thought of the last two points, good ideas.

You can link to the debugging doc there too https://github.com/DavidSpickett/ARMMultiTasking/blob/master/docs/debugging.md

File: docs/demos.md [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+* [alloc](../demos/alloc/README.md)
+* [loadbinary](../demos/loadbinary/README.md)
+* [loadbinaries](../demos/loadbinaries/README.md)
+* [loadpiebinary](../demos/loadpiebinary/README.md)
+* [parentchild](../demos/parentchild/README.md)
+* [permissions](../demos/permissions/README.md)
+* [trace](../demos/trace/README.md)
+* [signalhandling](../demos/signalhandling/README.md)
+* [backtrace](../demos/backtrace/README.md)
+* [fibres](../demos/fibres/README.md)
```
Comment (DavidSpickett) [inline docs/demos.md]: I'd rather you list just the ones that have been done. (assume that I'll merge this after spawn goes in)

**Review by DavidSpickett — CHANGES_REQUESTED** [review]
> By the way, should we put some brief description after each demo link?

There is a table in the main readme. I have been thinking about removing that. So what you could do is move that table into this doc and make the first column names link to the documents. (then all you have to do for a new doc is to add the link, the name/descriptions are already done for you)

Then add a "Features" section before "Building" in the main readme. That lists the top level features. I think this list would be:
* Cooperative multitasking
* Timer based thread switching
* Loading ELF programs, including position independent code
* Message passing
* Mutexes and condition variables
* Thread local storage
* Memory allocation
* File handling (via semihosting)
* Signal handling
* Permissions and errno for error handling
* Access to thread registers for tracing
* User space threads (fibres)
* Stack under/overflow detection

And anything you think I've missed. Keeping it to top level things. E.g. the ability to cancel a thread makes a good demo but I wouldn't list it as a top level "feature" since it's kinda implied.

File: docs/demos.md (hunk 3 above)
Comment (DavidSpickett) [reply docs/demos.md]: (but if you move the table here you can just link the first few names)

**Comment by OscarShiang** [comment]
All right, I have adjusted them.

**Comment by OscarShiang** [comment]
OK, I have adjusted them.

File: docs/demos.md [hunk 4]
```diff
@@ -0,0 +1,34 @@
+# AMT Demos
+
+The documents listed below give an overview of how the demos work and explain some mechanisms in these demos.
```
Comment (DavidSpickett) [inline docs/demos.md]: Reading it again, I think "linked" makes more sense than "listed".

File: README.md [hunk 5]
```diff
@@ -53,34 +69,6 @@ Each demo has `run_<demo>`, `debug_<demo>` and `test_<demo>` make targets.
 
 To test all the demos use `lit` (https://pypi.org/project/lit/) or just run `make`.
```
Comment (DavidSpickett) [inline README.md:71]: "For a full list see <link to demo doc>."

**Review by DavidSpickett — CHANGES_REQUESTED** [review]

**Comment by OscarShiang** [comment]
Alright, I updated it again.

**Review by DavidSpickett — APPROVED** [review]
Thanks!

**Pull request merged**